File: src/backend/java/org/example/Angular/entities/Category.java
package org.example.Angular.entities;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

@Entity
public class Category {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    private String name;
    private String image;

    @OneToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "category_subcategory",
            joinColumns = @JoinColumn(name = "category_id"),
            inverseJoinColumns = @JoinColumn(name = "subcategory_id"))
    private Set<SubCategory> subCategories = new HashSet<>();

//  Constructors

    public Category() { }

    public Category(String name) {
        this.name = name;
    }

    public Category(String name, String image) {
        this.name = name;
        this.image = image;
    }

//  Getters & Setters

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getImage() {
        return image;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public Set<SubCategory> getSubCategories() {
        return subCategories;
    }

    public void addSubCategory(SubCategory subcategory){
        subCategories.add(subcategory);
    }

    @Override
    public String toString() {
        return "Category{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", image='" + image + '\'' +
                ", subCategories=" + subCategories +
                '}';
    }
}

File: src/frontend/src/app/models/category.ts
import {Subcategory} from "./subcategory";

export class Category {
    private _id: number = 0;
    private _name: string = "";
    private _subcategories: Subcategory[] = [];
    private _image: string = "";

    constructor(id: number, name: string, subcategories: Subcategory[], image: string) {
        this._id = id;
        this._name = name;
        this._subcategories = subcategories;
        this._image = image;
    }

    get id(): number {
        return this._id;
    }

    set id(value: number) {
        this._id = value;
    }

    get name(): string {
        return this._name;
    }

    set name(value: string) {
        this._name = value;
    }

    get subcategories(): Subcategory[] {
        return this._subcategories;
    }

    set subcategories(value: Subcategory[]) {
        this._subcategories = value;
    }

    get image(): string {
        return this._image;
    }

    set image(value: string) {
        this._image = value;
    }
}